### PR TITLE
[Merged by Bors] - Config fragment/validation

### PR DESCRIPTION
**Pull request by teozkr**
## Description

This is an alternate take on #425. Essentially, we define a new macro `Fragment`, which creates a "parallel fragment universe" where everything is optional. For example, given the following types:

```rust
#[derive(Fragment)]
struct Outer {
    foo: u32,
    inner: Inner,
}

#[derive(Fragment)]
struct Inner {
    required: String,
    optional: Option<String>,
}
```

we generate the following fragment variants:

```rust
struct OuterFragment {
    foo: Option<u32>,
    inner: InnerFragment, // not an atomic type, so make every subfield optional instead
}

struct InnerFragment {
    required: Option<String>, // was non-optional, so make it optional
    optional: Option<String>, // was already optional, so preserve the type
}
```

as well as validation functions that turn `OuterFragment` into `Result<Outer>` (and an equivalent for `Inner`), while giving us the full context when a validation error occurs.

This allows us to safely deserialize and merge the fragment variants, and only afterwards validate that all required options are present in the final merged configuration.

This is a pretty big and breaking change, since it moves all `Serialize` and `Deserialize` impls on the resources into the fragment variants.

## Review Checklist
- [x] Code contains useful comments
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Comment by teozkr** [comment]
There is an example conversion of zookeeper-operator: https://github.com/stackabletech/zookeeper-operator/pull/530

**Review by adwk67 — COMMENTED** [review]
Some minor comments/questions aimed at clarifying things.

File: stackable-operator-derive/src/lib.rs
Comment (teozkr) [reply stackable-operator-derive/src/lib.rs:122]: Derivative provides alternate macros for deriving the std traits. You can't derive `Default` for the same type via both std's macro and Derivative.

File: stackable-operator-derive/src/lib.rs
Comment (teozkr) [reply stackable-operator-derive/src/lib.rs:73]: If `Bar` implements `FromFragment` manually then it will be whatever that impl defines `Bar::Fragment` to be. Otherwise the user will get a compile error complaining about `FromFragment` being required.

File: stackable-operator-derive/src/lib.rs
Comment (adwk67) [reply stackable-operator-derive/src/lib.rs:122]: then I'm a bit confused as we still have `#[derive(..., Default, ...)] ` in there: I thought that was referring to the std trait?

File: stackable-operator-derive/src/lib.rs
Comment (adwk67) [reply stackable-operator-derive/src/lib.rs:73]: Thanks for the clarification!

File: stackable-operator-derive/src/lib.rs
Comment (teozkr) [reply stackable-operator-derive/src/lib.rs:79]: This is used for managing `Option`-wrapping. If an atomic value (like `String`) is used in the full type then the fragment must make it `Option`al. If the field is already `Option<i32>` in the full type then we don't want it to be "double optional" (`Option<Option<i32>>`). Hence we use `RequiredFragment` when the container type already encodes this optionality.

File: src/commons/resources.rs
Comment (teozkr) [reply src/commons/resources.rs:116]: Not currently (we don't automatically derive ser/de either), but it might be worth checking whether it's required at all for our use case.

File: stackable-operator-derive/src/lib.rs
Comment (teozkr) [reply stackable-operator-derive/src/lib.rs:122]: Ah, that's for deriving the trait for the full type. `#[fragment_attrs(derive(Default))]` would break.

File: stackable-operator-derive/src/lib.rs
Comment (adwk67) [reply stackable-operator-derive/src/lib.rs:79]: Understood

File: src/commons/resources.rs
Comment (adwk67) [reply src/commons/resources.rs:116]: Ok

File: stackable-operator-derive/src/lib.rs
Comment (adwk67) [reply stackable-operator-derive/src/lib.rs:122]: Ok, thanks.

File: stackable-operator-derive/src/lib.rs
Comment (teozkr) [reply stackable-operator-derive/src/lib.rs:79]: There's probably a better way to phrase this in the inline docs…

File: CHANGELOG.md
Comment (teozkr) [reply CHANGELOG.md:17]: You should be able to get away with a mechanical search/replace, but I agree that it's not ideal. Originally I wanted to support this w/ a deprecation warning…

Comment (adwk67) [reply CHANGELOG.md:17]: ok, fair enough!

**Review by adwk67 — APPROVED** [review]

File: stackable-operator-derive/src/lib.rs
Comment (teozkr) [reply stackable-operator-derive/src/lib.rs:140]: ff53bdb6da12931714c4d140bf013b53d1ddcad0

File: stackable-operator-derive/src/lib.rs
Comment (teozkr) [reply stackable-operator-derive/src/lib.rs:79]: Tried to expand a bit in d38fb7b689ebbe9561a7c973a2c611ef5ca09c8a

**Comment by teozkr** [comment]
This depends on https://github.com/GREsau/schemars/pull/167

File: stackable-operator-derive/src/lib.rs
Comment (adwk67) [reply stackable-operator-derive/src/lib.rs:79]: Nice, thanks

**Review by adwk67 — APPROVED** [review]
lgtm.
(q:  should we refer to a tag rather than a branch - `feature/custom-bound` - for `schemars_derive`? Or is that going to remain static until the change is merged upstream?)

**Comment by teozkr** [comment]
>  (q: should we refer to a tag rather than a branch - feature/custom-bound - for schemars_derive?)

No, we should probably hold off on merging this until the branch is merged and released.

**Comment by adwk67** [comment]
> > (q: should we refer to a tag rather than a branch - feature/custom-bound - for schemars_derive?)
> 
> No, we should probably hold off on merging this until their branch is merged and released.

OK.

**Comment by teozkr** [comment]
@adwk67 Are we ok merging this now?

**Comment by adwk67** [comment]
> @adwk67 Are we ok merging this now?

yes, fine by me.

**Pull request closed without merging**
